File: J9Ch/src/J_7_Two_Pointers/Optional_11/_443_Two_Sum_Greater_than_target_Medium.cpp
/** 443 Two Sum - Greater than target*/


class Solution {
public:
    /**
     * @param nums: an array of integer
     * @param target: an integer
     * @return: an integer
     */
    int twoSum2(vector<int> &nums, int target) {
        // Write your code here
        int n = nums.size(), cnt = 0;
        sort(nums.begin(), nums.end());
        for (int i = 0; i < n; ++i) {
            cnt += find(i+1, n-1, nums, target-nums[i]);
        }
        return cnt;
    }
    int find(int l, int r, vector<int> &nums, int target) {
        int end = r;
        if (l > r)
            return 0;
        if (nums[l] > target)
            return r - l + 1;
        if (nums[r] <= target)
            return 0;
        int ans = r + 1;
        while (l <= r) {
            int mid = (l + r) >> 1;
            if (nums[mid] > target) {
                ans = mid;
                r = mid - 1;
            } else
                l = mid + 1;
        }
        return end - ans + 1;
    }
};

File: J9Ch/src/J_7_Two_Pointers/Optional_11/_443_Two_Sum_Greater_than_target_Medium.py
# /** 443 Two Sum - Greater than target


class Solution:
    # @param nums, an array of integer
    # @param target, an integer
    # @return an integer
    def twoSum2(self, nums, target):
        # Write your code here
        # print nums, target
        debug =False
        count = 0
        if (nums == None or len(nums) == 0):
            return count
        nums.sort()
        if debug:print nums
        i = 0
        j = len(nums) - 1
        while (i < j):
            if debug : print "new while"
            if debug:print "pos:",i,j
            if debug : print "value:",nums[i],nums[j]
            temp_sum = nums[i] + nums[j]
            if temp_sum <= target:
                # count +=len(nums)-1-j
                i += 1
            elif temp_sum > target:
                count+= j-i
                j -= 1
            if debug : print count
        if debug:print "len",len(nums)
        # if i!=len(nums)-2:count += (len(nums)-1-i + 1)*(len(nums)-1-i)/2
        return count